fe/fe-core/src/main/java/org/apache/doris/analysis/RoutineLoadDataSourceProperties.java

```diff
@@ -90,37 +146,242 @@ private void checkDataSourceProperties() throws AnalysisException {
+                throw new AnalysisException(CreateRoutineLoadStmt.KAFKA_DEFAULT_OFFSETS + " can only be set to OFFSET_BEGINNING, OFFSET_END or date time");
+            }
+            return false;
+        }
+    }
+
+    // init "kafkaPartitionOffsets" with partition property.
+    // The offset will be set to OFFSET_END for now, and will be changed in later analysis process.
+    private static void analyzeKafkaPartitionProperty(String kafkaPartitionsString,
+                                                      List<Pair<Integer, Long>> kafkaPartitionOffsets) throws AnalysisException {
+        kafkaPartitionsString = kafkaPartitionsString.replaceAll(" ", "");
+        if (kafkaPartitionsString.isEmpty()) {
+            throw new AnalysisException(CreateRoutineLoadStmt.KAFKA_PARTITIONS_PROPERTY + " could not be a empty string");
+        }
+        String[] kafkaPartitionsStringList = kafkaPartitionsString.split(",");
+        for (String s : kafkaPartitionsStringList) {
+            try {
+                kafkaPartitionOffsets.add(Pair.create(getIntegerValueFromString(s, CreateRoutineLoadStmt.KAFKA_PARTITIONS_PROPERTY),
+                        KafkaProgress.OFFSET_END_VAL));
+            } catch (AnalysisException e) {
+                throw new AnalysisException(CreateRoutineLoadStmt.KAFKA_PARTITIONS_PROPERTY
+                        + " must be a number string with comma-separated");
+            }
+        }
+    }
+
+    // Fill the partition's offset with given kafkaOffsetsString,
+    // Return true if offset is specified by timestamp.
+    private static boolean analyzeKafkaOffsetProperty(String kafkaOffsetsString, List<Pair<Integer, Long>> kafkaPartitionOffsets,
+                                                      String timeZoneStr)
+            throws UserException {
+        if (Strings.isNullOrEmpty(kafkaOffsetsString)) {
+            throw new AnalysisException(CreateRoutineLoadStmt.KAFKA_OFFSETS_PROPERTY + " could not be a empty string");
+        }
+        List<String> kafkaOffsetsStringList = Splitter.on(",").trimResults().splitToList(kafkaOffsetsString);
+        if (kafkaOffsetsStringList.size() != kafkaPartitionOffsets.size()) {
+            throw new AnalysisException("Partitions number should be equals to offsets number");
         }
 
-        // check custom properties
-        CreateRoutineLoadStmt.analyzeCustomProperties(properties, customKafkaProperties);
+        // We support two ways to specify the offset,
+        // one is to specify the offset directly, the other is to specify a timestamp.
+        // Doris will get the offset of the corresponding partition through the timestamp.
+        // The user can only choose one of these methods.
+        boolean foundTime = false;
+        boolean foundOffset = false;
+        for (String kafkaOffsetsStr : kafkaOffsetsStringList) {
+            if (TimeUtils.timeStringToLong(kafkaOffsetsStr) != -1) {
+                foundTime = true;
+            } else {
```

Comment:
else if (String is number) {}
else { format error}

Reply:
The kafkaOffsetsStr can be "OFFSET_END", "OFFSET_BEGINNING", number, "2020-10-10 00:00:00",
so only check if it is number is not enough.